### PR TITLE
statistics: trigger auto-analyze based on histogram row count

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/24237, related to https://github.com/pingcap/tidb/issues/26282

Problem Summary:

Trigger auto analyze in a more proper way.

### What is changed and how it works?

What's Changed:

Use the histogram row count as the base count for checking auto analyze, instead of the `Count` in `Table`.

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 

- Trigger auto-analyze based on histogram row count

### CODE REVIEW

File: statistics/handle/update_test.go [hunk 1]
```diff
@@ -2257,3 +2257,40 @@ func (s *testSerialStatsSuite) TestAutoUpdatePartitionInDynamicOnlyMode(c *C) {
...
+	tk.MustExec("explain select * from t where a = 1")
+	h := s.do.StatsHandle()
+	c.Assert(h.LoadNeededHistograms(), IsNil)
+	tk.MustExec(fmt.Sprintf("set global tidb_auto_analyze_start_time='00:00 +0000'"))
+	tk.MustExec(fmt.Sprintf("set global tidb_auto_analyze_end_time='23:59 +0000'"))
+
+	tk.MustExec("insert into t values (1)" + strings.Repeat(", (1)", 10))
+	c.Assert(s.do.StatsHandle().DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	c.Assert(h.Update(s.do.InfoSchema()), IsNil)
+	c.Assert(s.do.StatsHandle().HandleAutoAnalyze(s.do.InfoSchema()), IsTrue)
```
Comment (time-and-fate) [inline statistics/handle/update_test.go:2290]: It's a little strange that this assertion can pass on the current master branch.

File: statistics/handle/update_test.go [hunk 2]
```diff
@@ -2257,3 +2257,40 @@ func (s *testSerialStatsSuite) TestAutoUpdatePartitionInDynamicOnlyMode(c *C) {
...
+
+	tk.MustExec("delete from t limit 16")
+	c.Assert(s.do.StatsHandle().DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	c.Assert(h.Update(s.do.InfoSchema()), IsNil)
+	c.Assert(s.do.StatsHandle().HandleAutoAnalyze(s.do.InfoSchema()), IsTrue)
```
Comment (time-and-fate) [inline statistics/handle/update_test.go:2295]: I think auto analyze is triggered after deleting 16 lines from a 31 line table is expected. We should assert that auto analyze shouldn't be triggered after deleting 12 lines from this 31 line table.

**Review by time-and-fate — COMMENTED** [review]
I find that the newly added unit test can pass on the current master branch.

File: statistics/table.go [hunk 3]
```diff
@@ -214,6 +214,16 @@ func (t *Table) GetStatsInfo(ID int64, isIndex bool) (int64, *Histogram, *CMSket
 	return int64(colStatsInfo.TotalRowCount()), colStatsInfo.Histogram.Copy(), colStatsInfo.CMSketch.Copy(), colStatsInfo.TopN.Copy(), colStatsInfo.FMSketch.Copy()
 }
 
+// ColHistCount returns the count of the column histograms.
+func (t *Table) ColHistCount() float64 {
+	for _, col := range t.Columns {
+		if col != nil {
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply statistics/table.go:220]: Why do we only try to get row count from columns? I think indexes are also useful.

File: statistics/handle/update.go [hunk 4]
```diff
@@ -875,7 +875,11 @@ func NeedAnalyzeTable(tbl *statistics.Table, limit time.Duration, autoAnalyzeRat
...
 	// No need to analyze it.
-	if float64(tbl.ModifyCount)/float64(tbl.Count) <= autoAnalyzeRatio {
+	tblCnt := float64(tbl.Count)
+	if histCnt := tbl.ColHistCount(); histCnt > 0 {
+		tblCnt = histCnt
+	}
+	if float64(tbl.ModifyCount)/tblCnt <= autoAnalyzeRatio {
 		return false, ""
 	}
 	return true, fmt.Sprintf("too many modifications(%v/%v>%v)", tbl.ModifyCount, tbl.Count, autoAnalyzeRatio)
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply statistics/handle/update.go]: Done

File: statistics/handle/update_test.go (hunk 1 above)
Comment (eurekaka) [reply statistics/handle/update_test.go]: Fixed

File: statistics/handle/update_test.go (hunk 2 above)
Comment (eurekaka) [reply statistics/handle/update_test.go:2302]: Done

File: statistics/table.go [hunk 5]
```diff
@@ -214,6 +214,16 @@ func (t *Table) GetStatsInfo(ID int64, isIndex bool) (int64, *Histogram, *CMSket
 	return int64(colStatsInfo.TotalRowCount()), colStatsInfo.Histogram.Copy(), colStatsInfo.CMSketch.Copy(), colStatsInfo.TopN.Copy(), colStatsInfo.FMSketch.Copy()
 }
 
+// ColHistCount returns the count of the column histograms.
+func (t *Table) ColHistCount() float64 {
```
Comment (winoros) [inline statistics/table.go:218]: The name could be changed since the topn is split out of the histogram.

File: statistics/handle/update_test.go [hunk 6]
```diff
@@ -2284,3 +2284,47 @@ func (s *testSerialStatsSuite) TestAutoUpdatePartitionInDynamicOnlyMode(c *C) {
...
+
+	tk.MustExec("delete from t limit 12")
+	c.Assert(h.DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	c.Assert(h.Update(is), IsNil)
+	c.Assert(h.HandleAutoAnalyze(is), IsFalse)
+
+	tk.MustExec("delete from t limit 4")
+	c.Assert(s.do.StatsHandle().DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	c.Assert(h.Update(s.do.InfoSchema()), IsNil)
+	c.Assert(s.do.StatsHandle().HandleAutoAnalyze(s.do.InfoSchema()), IsTrue)
```
Comment (qw4990) [inline statistics/handle/update_test.go:2329]:
```suggestion
	c.Assert(h.DumpStatsDeltaToKV(handle.DumpAll), IsNil)
	c.Assert(h.Update(is), IsNil)
	c.Assert(h.HandleAutoAnalyze(s.do.InfoSchema()), IsTrue)
```

Comment (qw4990) [reply statistics/handle/update_test.go:2329]: LGTM; ping @eurekaka

File: statistics/handle/update_test.go [hunk 7]
```diff
@@ -2284,3 +2284,47 @@ func (s *testSerialStatsSuite) TestAutoUpdatePartitionInDynamicOnlyMode(c *C) {
...
+	tk.MustExec("insert into t values (1)" + strings.Repeat(", (1)", 19))
+	c.Assert(h.DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	is := s.do.InfoSchema()
+	c.Assert(h.Update(is), IsNil)
+	// To pass the stats.Pseudo check in autoAnalyzeTable
+	tk.MustExec("analyze table t")
+	tk.MustExec("explain select * from t where a = 1")
+	c.Assert(h.LoadNeededHistograms(), IsNil)
+	tk.MustExec(fmt.Sprintf("set global tidb_auto_analyze_start_time='00:00 +0000'"))
+	tk.MustExec(fmt.Sprintf("set global tidb_auto_analyze_end_time='23:59 +0000'"))
```
Comment (qw4990) [inline statistics/handle/update_test.go:2314]:
```suggestion
	tk.MustExec("set global tidb_auto_analyze_start_time='00:00 +0000'")
	tk.MustExec("set global tidb_auto_analyze_end_time='23:59 +0000'")
```

**Review by qw4990 — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**